src/main/java/de/qabel/desktop/storage/AbstractNavigation.java

```diff
@@ -0,0 +1,271 @@
+import javax.crypto.spec.SecretKeySpec;
+import java.io.*;
+import java.security.InvalidKeyException;
+import java.util.*;
+
+public abstract class AbstractNavigation implements BoxNavigation {
+
+	private static final Logger logger = LoggerFactory.getLogger(AbstractNavigation.class.getName());
+
+	DirectoryMetadata dm;
+	final QblECKeyPair keyPair;
+	final byte[] deviceId;
+	final CryptoUtils cryptoUtils;
+
+	final StorageReadBackend readBackend;
+	final StorageWriteBackend writeBackend;
+
+	private final Set<String> deleteQueue = new HashSet<>();
+	private final Set<FileUpdate> updatedFiles = new HashSet<>();
+
+
+
+	AbstractNavigation(DirectoryMetadata dm, QblECKeyPair keyPair, byte[] deviceId,
+	                   StorageReadBackend readBackend, StorageWriteBackend writeBackend) {
+		this.dm = dm;
+		this.keyPair = keyPair;
+		this.deviceId = deviceId;
+		this.readBackend = readBackend;
+		this.writeBackend = writeBackend;
+		cryptoUtils = new CryptoUtils();
+	}
+
+	@Override
+	public BoxNavigation navigate(BoxFolder target) throws QblStorageException {
+		try {
+			InputStream indexDl = readBackend.download(target.ref);
+			File tmp = File.createTempFile("dir", "db", dm.getTempDir());
+			KeyParameter key = new KeyParameter(target.key);
+			if (cryptoUtils.decryptFileAuthenticatedSymmetricAndValidateTag(indexDl, tmp, key)) {
+				DirectoryMetadata dm = DirectoryMetadata.openDatabase(
+						tmp, deviceId, target.ref, this.dm.getTempDir());
+				return new FolderNavigation(dm, keyPair, target.key, deviceId, readBackend, writeBackend);
+			} else {
+				throw new QblStorageNotFound("Invalid key");
+			}
+		} catch (IOException | InvalidKeyException e) {
+			throw new QblStorageException(e);
+		}
+	}
+
+	protected abstract DirectoryMetadata reloadMetadata() throws QblStorageException;
+
+	protected SecretKey makeKey(byte[] key2) {
+		return new SecretKeySpec(key2, "AES");
+	}
+
+	@Override
+	public void commit() throws QblStorageException {
+		byte[] version = dm.getVersion();
+		dm.commit();
+		logger.info("Committing version "+ new String(Hex.encodeHex(dm.getVersion()))
+				+ " with device id " + new String(Hex.encodeHex(dm.deviceId)));
+		DirectoryMetadata updatedDM = null;
+		try {
+			updatedDM = reloadMetadata();
+			logger.info("Remote version is " + new String(Hex.encodeHex(updatedDM.getVersion())));
+		} catch (QblStorageNotFound e) {
+			logger.info("Could not reload metadata");
+		}
+		// the remote version has changed from the _old_ version
+		if ((updatedDM != null) && (!Arrays.equals(version, updatedDM.getVersion()))) {
+			logger.info("Conflicting version");
+			// ignore our local directory metadata
+			// all changes that are not inserted in the new dm are _lost_!
+			dm = updatedDM;
+			for (FileUpdate update: updatedFiles) {
+				handleConflict(update);
+			}
+			dm.commit();
+		}
+		uploadDirectoryMetadata();
+		for (String ref: deleteQueue) {
+			writeBackend.delete(ref);
+		}
+		// TODO: make a test fail without these
```

Comment:
I am trying to, but the handleConflict() code us a bit too defensive and just doesn't want to fail.